### PR TITLE
New module: Do Not Disturb

### PR DESCRIPTION
**Pull request by maximbaz**
This module has a very simple idea behind it: to be able to turn on and off all the notifications in the system with a single click.

This is easy to achieve if all potentially distracting apps (like email, chat, etc.) use native system notifications rather than their own implementations. This module assumes such setup :wink: 

## Screenshots:

All the notifications are enabled:
image

Need to focus? A single mouse click disables the all system notifications:
image

Done working? A single mouse click, and you have all your notifications back:
image

## Implementation details

I couldn't figure out a smart way to disable notifications, that would work with all available notification managers. But I saw that many people suggest having a shell script that periodically kills the manager's process... So I took the same approach :smile: a bit ugly, but it works.

### CODE REVIEW

**Comment by tobes** [comment]
nice idea,

You can send signals to dunst

```
killall -SIGUSR1 dunst # pause
killall -SIGUSR2 dunst # resume
```

It does queue notifications which may not be what you wanted but `killall -SIGSTOP dunst` would kill the queue and re-enable

Now this would probably be dunst specific but as we have to define the notification process we could teach the module about new processes if needed.

**Comment by maximbaz** [comment]
Cool, I didn't know!

I think we should definitely _not_ queue notifications, with managers like `notify-osd` users would then sit for 2 hours seeing all old notifications, and the new ones will be blocked until all the old ones are gone :smile: 

Reading now about `pkill` vs `killall`: 
> `killall` by default it requires the precise process name, whereas `pkill` does basic pattern matching. Thus, `killall` is safer for users to blindly copy and paste.

I'm definitely switching to `killall` then!

I also tried `-SIGSTOP` vs `-SIGTERM` (which is default).
The `notify-osd` for example completely hangs after it receives the `-SIGSTOP` signal, moreover if there already was a visible notification on the screen, it stays visible forever. And if I execute `notify-send ...`, it will also hang.

I'd like to keep the module as simple and as generic as possible, i.e. not return to it for every new notification manager :smile:. Therefore I'd rather keep the simple `killall <process_name>`, which works for all. 

What do you think @tobes ?

**Comment by tobes** [comment]
> Therefore I'd rather keep the simple killall <process_name>, which works for all.

I don't like the fact that it is always running the thread.  You could probably adapt it so we only run the thread when we are in do not disturb mode, which would reduce the load.

The advantage of the specific approach is we really can cut down on the work done.  So maybe a compromise and we have a general `killall <process>` and we can also add specific managers as we see fit (but this can always be done later)

File: py3status/modules/do_not_disturb.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from threading import Thread, Event
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    format_off = 'OFF'
+    format_on = 'ON'
```
Comment (tobes) [inline py3status/modules/do_not_disturb.py]: I'd be tempted to have
```
format = '{state}'
state_on = 'ON'
state_off = 'OFF'
```

but maybe I'm a little insane

**Comment by maximbaz** [comment]
> I don't like the fact that it is always running the thread. You could probably adapt it so we only run the thread when we are in do not disturb mode, which would reduce the load.

The good thing is that thread is not running in an infinite loop when the "do not disturb" is disabled, the execution is stopped on this line and waiting for the event:

```
if self.running.wait():
```

I think such paused thread will consume almost no resources, and I'm not sure this is worse than creating and deleting threads on every click.

File: py3status/modules/do_not_disturb.py (hunk 1 above)
Comment (maximbaz) [inline py3status/modules/do_not_disturb.py]: Up to you, if you like it, I can change :wink: 

I took the current configuration from the example which I saw in my own `.i3status.conf`:

```
wireless wlp3s0 {
  format_up = ""
  format_down = ""
}
```

**Comment by tobes** [comment]
> if self.running.wait():

Yeah I saw that but after I made my comment.  Not noticed that before, always good to learn new things and I still have to take a look.

**Comment by maximbaz** [comment]
Just to confirm, is there any action from my side to be done before this pull request can be merged?

**Comment by tobes** [comment]
hmm,

I still like the state_.. over format_..

I'd also just not start the thread unless we actually need to.  Also I don't like the whole `kill ..` stuff constantly looping.  That just feels dirty.  I don't really like the sending USR1 to dunst either.  I can't find how dunst is actually being started on my system which is annoying.  I'd have thought this is where systemd might be useful to enable/disable services.

So that's where my thoughts are at.

**Comment by maximbaz** [comment]
While `state_` vs `format_` is an easy change, the other comments show a conceptual difference :wink:

I'm also not particularly fond of this approach, but I don't see any better ones for now, and since I use `notify-osd` these days I don't feel like implementing some special logic for `dunst` (that would for example use `SIGSTOP` instead of a thread with `SIGTERM` that works for all).

I'll keep on hosting this module locally then, and if you decide to merge it, ping me and I'll adapt the `format` parameter. Or feel free to take the code and adapt it to your preferences, if this module will still work fine with `notify-osd` I will gladly switch to use it :stuck_out_tongue_winking_eye:

**Comment by tobes** [comment]
I don't have a problem with it being merged as other users may find it useful.

I really like the concept and if your code works then that's a good start

File: py3status/modules/do_not_disturb.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                    if self.running.wait():
+                        system("killall '{}'".format(self.notification_manager))
+                        sleep(self.refresh_interval)
+
+        MyThread().start()
+
+    def do_not_disturb(self):
+        response = {
+            'cached_until': self.py3.CACHE_FOREVER,
+            'full_text': self.format.format(state=self.state_on if self.running.is_set() else self.state_off),
```
Comment (tobes) [inline py3status/modules/do_not_disturb.py]: this should really use `self.py3.safe_format()`

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,7 @@
 *.pyc
 build
 dist
+tags
```
Comment (ultrabug) [inline .gitignore]: I'm not sure I understand why you are modifying the gitignore mate ?

Comment (maximbaz-microsoft) [inline .gitignore]: that's an autogenerated ctags file, I used it at some point, but actually I can just revert this piece, doesn't matter

Comment (maximbaz) [inline .gitignore]: that's an autogenerated ctags file, I used it at some point, but actually I can just revert this piece, doesn't matter

**Comment by ultrabug** [comment]
Thanks a lot for your contribution once again @maximbaz !

**Pull request merged**